File: packages/python/tryaii/diagnose/cost.py
"""cost_exposure check (SPEC.md §2.4).

Pure arithmetic over the budget-module primitives; every number the check
cannot ground in the inventory or the other checks is itemized as
insufficient — never blended or guessed. Cache savings are an upper bound
built on conservative read-discount factors (shared/diagnose/costmodel.json).
"""

from __future__ import annotations

from typing import Any, Optional

from tryaii.budget import estimate_generation_cost, estimate_tokens

DAYS_PER_MONTH = 30
SWAP_QUALITY_EPSILON = 0.05
_CACHEABLE_VERDICTS = ("CACHEABLE", "CACHEABLE_WITH_ACTION", "CACHEABLE_PREFIX")


def _round4(x: float) -> float:
    return round(x, 4)


def _shell(reason: str, output_tokens: int,
           input_tokens: Optional[int] = None,
           token_method: Optional[str] = None) -> dict:
    return {
        "status": "insufficient_data",
        "reason": reason,
        "input_tokens": input_tokens,
        "output_tokens": output_tokens,
        "token_method": token_method,
        "cost_per_call_usd": None,
        "monthly": None,
        "swap": None,
    }


def run_cost(
    *,
    canonical_text: Optional[str],
    resolved_model_id: Optional[str],
    declared_model: Optional[str],
    registry,
    output_tokens: int,
    calls_per_day: Optional[float],
    cache_ctx: Optional[dict],
    fit_internal: Optional[dict],
    read_discount_factors: dict,
) -> dict:
    """cache_ctx: {verdict_code, stable_tokens, total_tokens, provider_key,
    upstream} extracted from the raw cachelint item (None when the cache
    check did not run). fit_internal: the model_fit internals (None when it
    did not run)."""

    # --- input tokens -----------------------------------------------------
    if cache_ctx is not None:
        input_tokens, token_method = cache_ctx["total_tokens"], "tokenizer"
    elif canonical_text is not None:
        input_tokens, token_method = estimate_tokens(canonical_text), "chars/4"
    else:
        return _shell("no prompt", output_tokens)

    # --- per-call cost ----------------------------------------------------
    if resolved_model_id is None:
        reason = (f"unknown model '{declared_model}'" if declared_model
                  else "no model declared")
        return _shell(reason, output_tokens, input_tokens, token_method)
    model = registry.get_model(resolved_model_id)
    cost_per_call = estimate_generation_cost(model, input_tokens, output_tokens)
    if cost_per_call is None:
        return _shell(f"no pricing for '{resolved_model_id}'",
                      output_tokens, input_tokens, token_method)
    cost_per_call = _round4(cost_per_call)

    # --- monthly ----------------------------------------------------------
    monthly: dict[str, Any]
    if calls_per_day is not None:
        monthly = {
            "status": "ok",
            "calls_per_day": calls_per_day,
            "cost_usd": _round4(cost_per_call * calls_per_day * DAYS_PER_MONTH),
        }
        savings = _cache_savings(cache_ctx, model, calls_per_day,
                                 read_discount_factors)
        if savings is not None:
            monthly["cache_savings_usd"] = savings
    else:
        monthly = {"status": "insufficient_data", "reason": "no traffic estimate"}

    # --- swap -------------------------------------------------------------
    swap = _swap(fit_internal, registry, input_tokens, output_tokens,
                 cost_per_call, calls_per_day)

    return {
        "status": "finding" if swap is not None else "ok",
        "reason": None,
        "input_tokens": input_tokens,
        "output_tokens": output_tokens,
        "token_method": token_method,
        "cost_per_call_usd": cost_per_call,
        "monthly": monthly,
        "swap": swap,
    }


def _cache_savings(cache_ctx: Optional[dict], model, calls_per_day: float,
                   factors: dict) -> Optional[float]:
    """Upper-bound monthly savings from prompt caching (SPEC.md §2.4)."""
    if cache_ctx is None or cache_ctx["verdict_code"] not in _CACHEABLE_VERDICTS:
        return None
    key = cache_ctx["provider_key"]
    factor = factors.get(key)
    if factor is None and key == "openrouter":
        factor = factors.get(cache_ctx.get("upstream") or "")
    if factor is None or not model or not model.pricing:
        return None
    return _round4(cache_ctx["stable_tokens"] / 1000
                   * model.pricing.input_per_1k
                   * factor * calls_per_day * DAYS_PER_MONTH)


def _swap(fit_internal: Optional[dict], registry, input_tokens: int,
          output_tokens: int, current_cost: float,
          calls_per_day: Optional[float]) -> Optional[dict]:
    """Cheapest ranked model within the quality tolerance (SPEC.md §2.4)."""
    if fit_internal is None or fit_internal.get("current_score") is None:
        return None
    floor = fit_internal["current_score"].quality_score - SWAP_QUALITY_EPSILON
    best: Optional[tuple[float, int, str]] = None  # (cost, rank, model_id)
    for rank0, score in enumerate(fit_internal["scores"]):
        if score.quality_score < floor:
            continue
        candidate = registry.get_model(score.model_id)
        cost = estimate_generation_cost(candidate, input_tokens, output_tokens)
        if cost is None:
            continue
        cost = _round4(cost)
        if best is None or cost < best[0]:  # ties break on rank (first wins)
            best = (cost, rank0 + 1, score.model_id)
    if best is None or best[0] >= current_cost:
        return None
    swap: dict[str, Any] = {"model_id": best[2], "cost_per_call_usd": best[0]}
    if calls_per_day is not None:
        swap["monthly_savings_usd"] = _round4(
            (current_cost - best[0]) * calls_per_day * DAYS_PER_MONTH)
    return swap

File: packages/node/src/diagnose/cost.ts
/**
 * cost_exposure check (SPEC.md §2.4).
 *
 * Pure arithmetic over the budget-module primitives; every number the check
 * cannot ground in the inventory or the other checks is itemized as
 * insufficient — never blended or guessed. Mirrors diagnose/cost.py.
 */

import { estimateGenerationCost } from '../budget.js';
import { halfEvenRound } from '../cachelint/util/halfEven.js';
import type { ModelRegistry } from '../registry/models.js';
import type { FitInternal } from './modelfit.js';

export const DAYS_PER_MONTH = 30;
export const SWAP_QUALITY_EPSILON = 0.05;
const CACHEABLE_VERDICTS = ['CACHEABLE', 'CACHEABLE_WITH_ACTION', 'CACHEABLE_PREFIX'];

export interface CacheCtx {
  verdict_code: string;
  stable_tokens: number;
  total_tokens: number;
  provider_key: string;
  upstream: string | null;
}

function round4(x: number): number {
  return halfEvenRound(x, 4);
}

/**
 * SPEC §2.4 token estimate: max(1, ceil(len/4)) with len in Unicode CODE
 * POINTS — Python's len(); NOT the UTF-16 text.length the budget module's
 * estimateTokens uses (they agree for BMP text).
 */
function estimateTokensCp(text: string): number {
  let cp = 0;
  for (const _ of text) cp++;
  return Math.max(1, Math.ceil(cp / 4));
}

function shell(
  reason: string,
  outputTokens: number,
  inputTokens: number | null = null,
  tokenMethod: string | null = null,
): Record<string, unknown> {
  return {
    status: 'insufficient_data',
    reason,
    input_tokens: inputTokens,
    output_tokens: outputTokens,
    token_method: tokenMethod,
    cost_per_call_usd: null,
    monthly: null,
    swap: null,
  };
}

export interface RunCostOptions {
  canonicalText: string | null;
  resolvedModelId: string | null;
  declaredModel: string | null;
  registry: ModelRegistry;
  outputTokens: number;
  callsPerDay: number | null;
  cacheCtx: CacheCtx | null;
  fitInternal: FitInternal | null;
  readDiscountFactors: Record<string, number | null>;
}

export function runCost(opts: RunCostOptions): Record<string, unknown> {
  const {
    canonicalText,
    resolvedModelId,
    declaredModel,
    registry,
    outputTokens,
    callsPerDay,
    cacheCtx,
    fitInternal,
    readDiscountFactors,
  } = opts;

  // --- input tokens -----------------------------------------------------
  let inputTokens: number;
  let tokenMethod: string;
  if (cacheCtx !== null) {
    inputTokens = cacheCtx.total_tokens;
    tokenMethod = 'tokenizer';
  } else if (canonicalText !== null) {
    inputTokens = estimateTokensCp(canonicalText);
    tokenMethod = 'chars/4';
  } else {
    return shell('no prompt', outputTokens);
  }

  // --- per-call cost ----------------------------------------------------
  if (resolvedModelId === null) {
    const reason = declaredModel ? `unknown model '${declaredModel}'` : 'no model declared';
    return shell(reason, outputTokens, inputTokens, tokenMethod);
  }
  const model = registry.getModel(resolvedModelId);
  const rawCost = estimateGenerationCost(model, inputTokens, outputTokens);
  if (rawCost === null) {
    return shell(`no pricing for '${resolvedModelId}'`, outputTokens, inputTokens, tokenMethod);
  }
  const costPerCall = round4(rawCost);

  // --- monthly ----------------------------------------------------------
  let monthly: Record<string, unknown>;
  if (callsPerDay !== null) {
    monthly = {
      status: 'ok',
      calls_per_day: callsPerDay,
      cost_usd: round4(costPerCall * callsPerDay * DAYS_PER_MONTH),
    };
    const savings = cacheSavings(cacheCtx, resolvedModelId, registry, callsPerDay, readDiscountFactors);
    if (savings !== null) monthly.cache_savings_usd = savings;
  } else {
    monthly = { status: 'insufficient_data', reason: 'no traffic estimate' };
  }

  // --- swap -------------------------------------------------------------
  const swap = findSwap(fitInternal, registry, inputTokens, outputTokens, costPerCall, callsPerDay);

  return {
    status: swap !== null ? 'finding' : 'ok',
    reason: null,
    input_tokens: inputTokens,
    output_tokens: outputTokens,
    token_method: tokenMethod,
    cost_per_call_usd: costPerCall,
    monthly,
    swap,
  };
}

/** Upper-bound monthly savings from prompt caching (SPEC.md §2.4). */
function cacheSavings(
  cacheCtx: CacheCtx | null,
  resolvedModelId: string,
  registry: ModelRegistry,
  callsPerDay: number,
  factors: Record<string, number | null>,
): number | null {
  if (cacheCtx === null || !CACHEABLE_VERDICTS.includes(cacheCtx.verdict_code)) return null;
  const key = cacheCtx.provider_key;
  let factor = factors[key] ?? null;
  if (factor === null && key === 'openrouter') {
    factor = factors[cacheCtx.upstream ?? ''] ?? null;
  }
  const model = registry.getModel(resolvedModelId);
  if (factor === null || !model || !model.pricing) return null;
  return round4(
    (cacheCtx.stable_tokens / 1000) *
      model.pricing.inputPer1k *
      factor *
      callsPerDay *
      DAYS_PER_MONTH,
  );
}

/** Cheapest ranked model within the quality tolerance (SPEC.md §2.4). */
function findSwap(
  fitInternal: FitInternal | null,
  registry: ModelRegistry,
  inputTokens: number,
  outputTokens: number,
  currentCost: number,
  callsPerDay: number | null,
): Record<string, unknown> | null {
  if (fitInternal === null || fitInternal.currentScore === null) return null;
  const floor = fitInternal.currentScore.qualityScore - SWAP_QUALITY_EPSILON;
  let best: { cost: number; rank: number; modelId: string } | null = null;
  for (let rank0 = 0; rank0 < fitInternal.scores.length; rank0++) {
    const score = fitInternal.scores[rank0];
    if (score.qualityScore < floor) continue;
    const candidate = registry.getModel(score.modelId);
    const rawCost = estimateGenerationCost(candidate, inputTokens, outputTokens);
    if (rawCost === null) continue;
    const cost = round4(rawCost);
    if (best === null || cost < best.cost) {
      // ties break on rank (first wins)
      best = { cost, rank: rank0 + 1, modelId: score.modelId };
    }
  }
  if (best === null || best.cost >= currentCost) return null;
  const swap: Record<string, unknown> = {
    model_id: best.modelId,
    cost_per_call_usd: best.cost,
  };
  if (callsPerDay !== null) {
    swap.monthly_savings_usd = round4((currentCost - best.cost) * callsPerDay * DAYS_PER_MONTH);
  }
  return swap;
}
